File: android/deeplinkSDK/src/main/kotlin/com/deeplink/sdk/ApiClient.kt
package com.deeplink.sdk

import android.os.Build
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

internal object ApiClient {

    fun fetchInitData(
        config: DeeplinkConfig,
        deviceSignals: Map<String, String?> = emptyMap(),
        referrerClickId: String? = null,
        callback: (DeeplinkData?) -> Unit,
    ) {
        Thread {
            val body = JSONObject().apply {
                put("api_key", config.apiKey)
                put("user_agent", buildUserAgent())
                deviceSignals.forEach { (k, v) -> if (v != null) put(k, v) }
                if (referrerClickId != null) {
                    put("referrer_click_id", referrerClickId)
                    DeeplinkLogger.log("fetchInitData — referrer_click_id=$referrerClickId (deterministic)")
                }
            }
            DeeplinkLogger.log("fetchInitData → POST /sdk/init signals=${deviceSignals.keys}")
            val result = post(config, "/sdk/init", body, timeout = 10_000) { json ->
                val matched = json.optBoolean("matched", false)
                DeeplinkLogger.log("fetchInitData ← matched=$matched")
                if (!matched) return@post null
                val data = json.optJSONObject("data") ?: return@post null
                val meta = mutableMapOf<String, String>()
                data.optJSONObject("metadata")?.let { m ->
                    m.keys().forEach { k -> meta[k] = m.optString(k) }
                }
                DeeplinkData(
                    linkId         = data.getString("linkId"),
                    alias          = data.getString("alias"),
                    iosUrl         = data.optString("iosUrl").takeIf { it.isNotEmpty() },
                    androidUrl     = data.optString("androidUrl").takeIf { it.isNotEmpty() },
                    destinationUrl = data.getString("destinationUrl"),
                    utmSource      = data.optString("utmSource").takeIf { it.isNotEmpty() },
                    utmMedium      = data.optString("utmMedium").takeIf { it.isNotEmpty() },
                    utmCampaign    = data.optString("utmCampaign").takeIf { it.isNotEmpty() },
                    utmContent     = data.optString("utmContent").takeIf { it.isNotEmpty() },
                    utmTerm        = data.optString("utmTerm").takeIf { it.isNotEmpty() },
                    metadata       = meta,
                    creativeName   = data.optString("creativeName").takeIf { it.isNotEmpty() },
                    creativeId     = data.optString("creativeId").takeIf { it.isNotEmpty() },
                )
            }
            callback(result)
        }.start()
    }

    fun createLink(
        config: DeeplinkConfig,
        destination: String,
        params: Map<String, String>,
        iosUrl: String? = null,
        androidUrl: String? = null,
        alias: String? = null,
        title: String? = null,
        description: String? = null,
        ogImage: String? = null,
        utmSource: String? = null,
        utmMedium: String? = null,
        utmCampaign: String? = null,
        utmContent: String? = null,
        utmTerm: String? = null,
        expiresAt: String? = null,
        callback: (CreatedLink?) -> Unit,
    ) {
        Thread {
            val body = JSONObject().apply {
                put("api_key", config.apiKey)
                put("destination_url", destination)
                if (params.isNotEmpty()) put("params", JSONObject(params))
                iosUrl?.let      { put("ios_url", it) }
                androidUrl?.let  { put("android_url", it) }
                alias?.let       { put("alias", it) }
                title?.let       { put("title", it) }
                description?.let { put("description", it) }
                ogImage?.let     { put("og_image", it) }
                utmSource?.let   { put("utm_source", it) }
                utmMedium?.let   { put("utm_medium", it) }
                utmCampaign?.let { put("utm_campaign", it) }
                utmContent?.let  { put("utm_content", it) }
                utmTerm?.let     { put("utm_term", it) }
                expiresAt?.let   { put("expires_at", it) }
            }
            DeeplinkLogger.log("createLink → dest=$destination")
            val result = post(config, "/sdk/link", body) { json ->
                DeeplinkLogger.log("createLink ← url=${json.optString("url")}")
                CreatedLink(
                    url    = json.getString("url"),
                    alias  = json.getString("alias"),
                    linkId = json.getString("link_id"),
                )
            }
            callback(result)
        }.start()
    }

    fun fetchLinkData(config: DeeplinkConfig, alias: String, callback: (DeeplinkData?) -> Unit) {
        Thread {
            DeeplinkLogger.log("fetchLinkData → alias=$alias")
            val result = get(config, "/sdk/resolve/$alias") { json ->
                val matched = json.optBoolean("matched", false)
                if (!matched) return@get null
                val data = json.optJSONObject("data") ?: return@get null
                val meta = mutableMapOf<String, String>()
                data.optJSONObject("metadata")?.let { m ->
                    m.keys().forEach { k -> meta[k] = m.optString(k) }
                }
                DeeplinkData(
                    linkId         = data.getString("linkId"),
                    alias          = data.getString("alias"),
                    iosUrl         = data.optString("iosUrl").takeIf { it.isNotEmpty() },
                    androidUrl     = data.optString("androidUrl").takeIf { it.isNotEmpty() },
                    destinationUrl = data.getString("destinationUrl"),
                    utmSource      = data.optString("utmSource").takeIf { it.isNotEmpty() },
                    utmMedium      = data.optString("utmMedium").takeIf { it.isNotEmpty() },
                    utmCampaign    = data.optString("utmCampaign").takeIf { it.isNotEmpty() },
                    utmContent     = data.optString("utmContent").takeIf { it.isNotEmpty() },
                    utmTerm        = data.optString("utmTerm").takeIf { it.isNotEmpty() },
                    metadata       = meta,
                    creativeName   = data.optString("creativeName").takeIf { it.isNotEmpty() },
                    creativeId     = data.optString("creativeId").takeIf { it.isNotEmpty() },
                )
            }
            DeeplinkLogger.log("fetchLinkData ← metadata=${result?.metadata}")
            callback(result)
        }.start()
    }

    fun recordImpression(config: DeeplinkConfig, alias: String, callback: ((Boolean) -> Unit)? = null) {
        Thread {
            val body = JSONObject().apply {
                put("api_key", config.apiKey)
                put("link_alias", alias)
                put("platform", "android")
            }
            post(config, "/api/impressions", body) { _: JSONObject -> null }
            android.os.Handler(android.os.Looper.getMainLooper()).post { callback?.invoke(true) }
        }.start()
    }

    fun trackEvent(config: DeeplinkConfig, name: String, properties: Map<String, Any>, sessionId: String) {
        Thread {
            val body = JSONObject().apply {
                put("api_key", config.apiKey)
                put("name", name)
                put("session_id", sessionId)
                if (properties.isNotEmpty()) put("properties", JSONObject(properties))
            }
            DeeplinkLogger.log("track '$name' props=$properties")
            post(config, "/api/events", body) { _: JSONObject -> null }
        }.start()
    }

    // MARK: - HTTP helper (synchronous — always call from a background thread)

    private fun <T> post(
        config: DeeplinkConfig,
        path: String,
        body: JSONObject,
        timeout: Int = 15_000,
        parse: (JSONObject) -> T?,
    ): T? = try {
        val conn = (URL("${config.apiBaseUrl}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json")
            doOutput = true
            connectTimeout = timeout
            readTimeout = timeout
        }
        conn.outputStream.use { it.write(body.toString().toByteArray()) }
        if (conn.responseCode !in 200..299) {
            DeeplinkLogger.error("POST $path returned ${conn.responseCode}")
            return null
        }
        parse(JSONObject(conn.inputStream.bufferedReader().readText()))
    } catch (e: Exception) {
        DeeplinkLogger.error("POST $path failed", e)
        null
    }

    private fun <T> get(
        config: DeeplinkConfig,
        path: String,
        timeout: Int = 10_000,
        parse: (JSONObject) -> T?,
    ): T? = try {
        val conn = (URL("${config.apiBaseUrl}$path?api_key=${config.apiKey}").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = timeout
            readTimeout = timeout
        }
        if (conn.responseCode !in 200..299) {
            DeeplinkLogger.error("GET $path returned ${conn.responseCode}")
            return null
        }
        parse(JSONObject(conn.inputStream.bufferedReader().readText()))
    } catch (e: Exception) {
        DeeplinkLogger.error("GET $path failed", e)
        null
    }

    private fun buildUserAgent(): String =
        "DeeplinkSDK-Android/${Build.MODEL} Android/${Build.VERSION.RELEASE}"
}

File: ios/Sources/DeeplinkSDK/APIClient.swift
import Foundation
import UIKit

internal final class APIClient {
    private let config: DeeplinkConfig
    private let session: URLSession

    /// Set to true by Deeplink.checkPasteboardOnInstall()
    var pasteboardCheckEnabled: Bool = false

    /// In-flight deduplication — queues completions so calling getInitData() and
    /// onFirstLaunch() together in AppDelegate fires only ONE network request.
    private var pendingInitCompletions: [(DeeplinkData?) -> Void] = []
    private var initFetchInFlight = false
    private let initLock = NSLock()

    init(config: DeeplinkConfig) {
        self.config = config
        self.session = URLSession(configuration: .default)
    }

    // MARK: - SDK Endpoints

    func fetchInitData(completion: @escaping (DeeplinkData?) -> Void) {
        // Collect all UIKit signals on the main thread first.
        // UIScreen.main, UIDevice, UIPasteboard must be accessed on main thread.
        // Then fire the network call on a background queue — never block main.
        let collectAndFire = {
            var body: [String: Any] = [
                "api_key":    self.config.apiKey,
                "user_agent": self.userAgent(),
            ]

            // Keychain device ID — survives reinstall, deterministic on returning users.
            let deviceId = KeychainHelper.getOrCreateDeviceId()
            body["device_id"] = deviceId
            DeeplinkLogger.log("fetchInitData — device_id: \(deviceId)")

            // IDFV — Apple's stable per-vendor identifier, no ATT required.
            if let idfv = UIDevice.current.identifierForVendor?.uuidString {
                body["idfv"] = idfv
                DeeplinkLogger.log("fetchInitData — idfv: \(idfv)")
            }

            // Clipboard attribution (opt-in via checkPasteboardOnInstall).
            // Note: reading UIPasteboard.general shows the iOS 16+ "pasted from" toast.
            if self.pasteboardCheckEnabled {
                if let pasteStr = UIPasteboard.general.string,
                   pasteStr.hasPrefix("deeplink-click:") {
                    let clickId = String(pasteStr.dropFirst("deeplink-click:".count))
                    if !clickId.isEmpty {
                        body["pasteboard_click_id"] = clickId
                        UIPasteboard.general.string = nil
                        DeeplinkLogger.log("fetchInitData — clipboard click_id found: \(clickId)")
                    }
                }
            }

            // Native device signals for probabilistic matching
            let model = self.deviceModel()
            let osVer = UIDevice.current.systemVersion
            let screen = self.screenRes()
            body["device_model"] = model
            body["os_version"]   = osVer
            body["screen_res"]   = screen
            body["timezone"]     = TimeZone.current.identifier
            body["language"]     = Locale.current.languageCode ?? "en"

            DeeplinkLogger.log("fetchInitData — signals: model=\(model) os=\(osVer) screen=\(screen)")

            // Deduplicate: if a fetch is already in flight, queue and return
            self.initLock.lock()
            self.pendingInitCompletions.append(completion)
            let shouldFire = !self.initFetchInFlight
            if shouldFire { self.initFetchInFlight = true }
            self.initLock.unlock()

            guard shouldFire else {
                DeeplinkLogger.log("fetchInitData — request already in flight, queued")
                return
            }

            // Network call on background thread
            DispatchQueue.global(qos: .userInitiated).async {
                self.post("/sdk/init", body: body) { (response: SDKInitResponse?) in
                    if let response {
                        DeeplinkLogger.log("fetchInitData — matched=\(response.matched) alias=\(response.data?.alias ?? "none")")
                    }
                    let data = (response?.matched == true) ? response?.data?.toDeeplinkData() : nil

                    // Drain all queued completions with the same result
                    self.initLock.lock()
                    let completions = self.pendingInitCompletions
                    self.pendingInitCompletions = []
                    self.initFetchInFlight = false
                    self.initLock.unlock()

                    completions.forEach { $0(data) }
                }
            }
        }

        if Thread.isMainThread {
            collectAndFire()
        } else {
            DispatchQueue.main.async { collectAndFire() }
        }
    }

    func createLink(
        destination: String,
        params: [String: String],
        iosUrl: String? = nil,
        androidUrl: String? = nil,
        alias: String? = nil,
        title: String? = nil,
        description: String? = nil,
        ogImage: String? = nil,
        utmSource: String? = nil,
        utmMedium: String? = nil,
        utmCampaign: String? = nil,
        utmContent: String? = nil,
        utmTerm: String? = nil,
        expiresAt: String? = nil,
        completion: @escaping (CreatedLink?, Error?) -> Void
    ) {
        var body: [String: Any] = [
            "api_key": config.apiKey,
            "destination_url": destination,
        ]
        if !params.isEmpty     { body["params"]        = params }
        if let v = iosUrl      { body["ios_url"]       = v }
        if let v = androidUrl  { body["android_url"]   = v }
        if let v = alias       { body["alias"]         = v }
        if let v = title       { body["title"]         = v }
        if let v = description { body["description"]   = v }
        if let v = ogImage     { body["og_image"]      = v }
        if let v = utmSource   { body["utm_source"]    = v }
        if let v = utmMedium   { body["utm_medium"]    = v }
        if let v = utmCampaign { body["utm_campaign"]  = v }
        if let v = utmContent  { body["utm_content"]   = v }
        if let v = utmTerm     { body["utm_term"]      = v }
        if let v = expiresAt   { body["expires_at"]    = v }

        DeeplinkLogger.log("createLink — dest=\(destination)")

        post("/sdk/link", body: body) { (parsed: SDKCreateLinkResponse?) in
            guard let parsed else {
                completion(nil, URLError(.cannotParseResponse)); return
            }
            DeeplinkLogger.log("createLink — url=\(parsed.url)")
            completion(CreatedLink(url: parsed.url, alias: parsed.alias, linkId: parsed.linkId), nil)
        }
    }

    func fetchLinkData(alias: String, completion: @escaping (DeeplinkData?) -> Void) {
        guard let url = URL(string: config.apiBaseURL.absoluteString + "/sdk/resolve/\(alias)?api_key=\(config.apiKey)")
        else { completion(nil); return }

        DeeplinkLogger.log("fetchLinkData — alias=\(alias)")

        var req = URLRequest(url: url, timeoutInterval: 10)
        req.httpMethod = "GET"

        session.dataTask(with: req) { data, response, error in
            if let error { DeeplinkLogger.error("fetchLinkData error: \(error)"); completion(nil); return }
            guard let data,
                  let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode)
            else { completion(nil); return }

            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            if let resp = try? decoder.decode(SDKInitResponse.self, from: data), resp.matched {
                DeeplinkLogger.log("fetchLinkData — metadata=\(resp.data?.metadata ?? [:])")
                completion(resp.data?.toDeeplinkData())
            } else {
                completion(nil)
            }
        }.resume()
    }

    func recordImpression(alias: String, completion: ((Bool) -> Void)? = nil) {
        post("/api/impressions", body: [
            "api_key": config.apiKey,
            "link_alias": alias,
            "platform": "ios",
        ]) { (_: EmptyResponse?) in completion?(true) }
    }

    func trackEvent(name: String, properties: [String: Any], sessionId: String, completion: ((Bool) -> Void)? = nil) {
        var body: [String: Any] = [
            "api_key": config.apiKey,
            "name": name,
            "session_id": sessionId,
        ]
        if !properties.isEmpty { body["properties"] = properties }
        DeeplinkLogger.log("track event '\(name)' properties=\(properties)")
        post("/api/events", body: body) { (_: EmptyResponse?) in completion?(true) }
    }

    // MARK: - Common HTTP helper

    private func post<T: Decodable>(_ path: String, body: [String: Any], completion: @escaping (T?) -> Void) {
        guard
            let url = URL(string: config.apiBaseURL.absoluteString + path),
            let bodyData = try? JSONSerialization.data(withJSONObject: body)
        else { completion(nil); return }

        var req = URLRequest(url: url, timeoutInterval: 15)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = bodyData

        session.dataTask(with: req) { data, response, error in
            if let error { DeeplinkLogger.error("POST \(path) error: \(error)") }
            guard
                let data,
                let http = response as? HTTPURLResponse,
                (200...299).contains(http.statusCode)
            else { completion(nil); return }

            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            completion(try? decoder.decode(T.self, from: data))
        }.resume()
    }

    // MARK: - Helpers

    private func userAgent() -> String {
        let info = Bundle.main.infoDictionary
        let appName    = info?["CFBundleName"] as? String ?? "App"
        let appVersion = info?["CFBundleShortVersionString"] as? String ?? "1.0"
        let os = "iOS \(ProcessInfo.processInfo.operatingSystemVersionString)"
        return "\(appName)/\(appVersion) \(os)"
    }

    /// Thread-safe hardware model via sysctlbyname, e.g. "iPhone15,2".
    private func deviceModel() -> String {
        var size = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        var machine = [CChar](repeating: 0, count: size)
        sysctlbyname("hw.machine", &machine, &size, nil, 0)
        return String(cString: machine)
    }

    /// Must be called on the main thread — enforced by fetchInitData().
    private func screenRes() -> String {
        let s = UIScreen.main
        let w = Int(s.bounds.width)
        let h = Int(s.bounds.height)
        let scale = Int(s.scale)
        return "\(w)x\(h)x\(scale)"
    }
}

private struct EmptyResponse: Decodable {}
